firedrake/assign.py

```diff
@@ -0,0 +1,130 @@
+import abc
+
+import numpy as np
+from pyop2.utils import cached_property
+import ufl
+from ufl.algorithms import extract_coefficients
+from ufl.constantvalue import RealValue, Zero
+from ufl.corealg.map_dag import map_expr_dag
+from ufl.corealg.multifunction import MultiFunction
+
+from firedrake.utils import ScalarType
+from firedrake.vector import Vector
+
+
+class WeightedCoefficientCollector(MultiFunction):
+    def __init__(self):
+        super().__init__()
+        self.weighted_coefficients = []
+
+    def sum(self, o):
+        for op in o.ufl_operands:
+            self(op)
```

Comment:
Don't hand traverse, instead write:
```
def sum(self, o, *operands):
    pass
```

Hand-traversal destroys the DAG-based visitor pattern with the result that you will run slower than you want.